### PR TITLE
Interface Relationship: Root level matching and singleton relationship bugfix

### PR DESCRIPTION
**Pull request by maxwellwynter**
# Description
This fork addresses two problems with the recent interface relationship feature.

1.Applies a fix to the error discussed in issue #585
2.Adds the ability to filter interface implementations on a parent nodes where clause.

e.g
```graphql
actors(where:{actedInConnection:{node:{title:"The Matrix", _on:{Movie:{title:"Matrix"}}}}}) {
                    actedIn {
                        title
                        ... on Movie {
                            runtime
                        }
                        ... on Series {
                            episodes
                        }
                    }
                }
```

# Issue

The library currently documents connection where attributes for nodes with relationship interfaces, though they are not implemented and no WHERE cypher is generated. This simply adds the missing feature.

# Checklist

The following requirements should have been met (depending on the changes in the branch):

- [ ] Documentation has been updated
- [ ] TCK tests have been updated
- [ ] Integration tests have been updated
- [ ] Example applications have been updated
- [ ] New files have copyright header
- [ ] CLA (https://neo4j.com/developer/cla/) has been signed

Fixes: #585

### CODE REVIEW

**Comment by darrellwarde** [comment]
Thanks @maxwellprioritynetworks! I could have sworn you had signed the CLA before but I must be mistaken. Can you please follow the steps at https://neo4j.com/developer/cla/#sign-cla ? Thanks!

**Comment by maxwellprioritynetworks** [comment]
done.

File: packages/graphql/src/translate/create-projection-and-params.ts [hunk 1]
```diff
@@ -338,8 +338,7 @@ function createProjectionAndParams({
                 if (optionsInput) {
                     offsetLimitStr = createOffsetLimitStr({ offset: optionsInput.offset, limit: optionsInput.limit });
                 }
-
-                res.projection.push(`${f.alias}: collect(${f.alias})${offsetLimitStr}`);
+                res.projection.push(`${f.alias}: ${!isArray ? "head(" : ""}collect(${f.alias})${offsetLimitStr}${!isArray ? ")" : ""}`);
```
Comment (darrellwarde) [inline packages/graphql/src/translate/create-projection-and-params.ts:341]: Really need to be seeing an integration test or 2 to ensure that this fix works.

File: packages/graphql/tests/tck/tck-test-files/interface-relationships/read.test.ts [hunk 2]
```diff
@@ -59,6 +59,63 @@ describe("Interface Relationships", () => {
...
+                        ... on Movie {
+                            runtime
+                        }
+                        ... on Series {
+                            episodes
+                        }
+                    }
+                }
+            }
+        `;
```
Comment (darrellwarde) [inline packages/graphql/tests/tck/tck-test-files/interface-relationships/read.test.ts:77]: Can you please make sure that formatting has been applied? This query should look a lot nicer.

**Review by darrellwarde — CHANGES_REQUESTED** [review]
Sorry for the delay in getting this reviewed. We really need to see some integration tests, both for the singleton fix and for the new filtering feature, before we can merge this in. Hopefully you can find the time to look at this!

File: packages/graphql/tests/tck/tck-test-files/interface-relationships/read.test.ts (hunk 2 above)
Comment (maxwellprioritynetworks) [reply packages/graphql/tests/tck/tck-test-files/interface-relationships/read.test.ts]: done

**Comment by darrellwarde** [comment]
Thanks for adding the integration test for the singleton fix. Apologies if you're working on it already, but if we could just get one integration test validating a query like the below, then this will be good to be merged and will be released early next week!
```graphql
query {
    actors(where : { actedInConnection: { node : { title : "The Matrix" , _on : { Movie : { title : "Matrix" } } } } }) {
         actedIn {
             title
             ... on Movie {
                 runtime
             }
             ... on Series {
                 episodes
             }
         }
     }
 }
```

**Comment by maxwellprioritynetworks** [comment]
Hi Darrell, Ill have to do this over the weekend. Its a busy week for me.

**Comment by darrellwarde** [comment]
> Hi Darrell, Ill have to do this over the weekend. Its a busy week for me.

Of course, no stress and no pressure! 🧘

**Comment by darrellwarde** [comment]
Hey @maxwellprioritynetworks, I'm having a look at PRs today and have come around to this. I had a look at adding an integration test, and added the following:

```ts
test("should read and return interface relationship fields with where override when filtering from top-level query", async () => {
        const session = driver.session();

        const actorName = generate({
            readable: true,
            charset: "alphabetic",
        });

        const movieTitle = generate({
            readable: true,
            charset: "alphabetic",
        });
        const movieRuntime = faker.random.number();
        const movieScreenTime = faker.random.number();

        const movieTitleNonMatch = generate({
            readable: true,
            charset: "alphabetic",
        });
        const movieRuntimeNonMatch = faker.random.number();
        const movieScreenTimeNonMatch = faker.random.number();

        const seriesTitle = generate({
            readable: true,
            charset: "alphabetic",
        });
        const seriesEpisodes = faker.random.number();
        const seriesScreenTime = faker.random.number();

        const seriesTitleNonMatch = generate({
            readable: true,
            charset: "alphabetic",
        });
        const seriesEpisodesNonMatch = faker.random.number();
        const seriesScreenTimeNonMatch = faker.random.number();

        const query = `
            query {
                actors(
                    where: { actedInConnection: { node: { title: "${seriesTitle}", _on: { Movie: { title: "${movieTitle}" } } } } }
                ) {
                    name
                    actedIn {
                        title
                        ... on Movie {
                            runtime
                        }
                        ... on Series {
                            episodes
                        }
                    }
                }
            }
        `;

        try {
            await session.run(
                `
                CREATE (a:Actor { name: $actorName })
                CREATE (a)-[:ACTED_IN { screenTime: $movieScreenTime }]->(:Movie { title: $movieTitle, runtime:$movieRuntime })
                CREATE (a)-[:ACTED_IN { screenTime: $seriesScreenTime }]->(:Series { title: $seriesTitle, episodes: $seriesEpisodes })
                CREATE (a)-[:ACTED_IN { screenTime: $movieScreenTimeNonMatch }]->(:Movie { title: $movieTitleNonMatch, runtime:$movieRuntimeNonMatch })
                CREATE (a)-[:ACTED_IN { screenTime: $seriesScreenTimeNonMatch }]->(:Series { title: $seriesTitleNonMatch, episodes: $seriesEpisodesNonMatch })
            `,
                {
                    actorName,
                    movieTitle,
                    movieRuntime,
                    movieScreenTime,
                    seriesTitle,
                    seriesEpisodes,
                    seriesScreenTime,
                    movieTitleNonMatch,
                    movieRuntimeNonMatch,
                    movieScreenTimeNonMatch,
                    seriesTitleNonMatch,
                    seriesEpisodesNonMatch,
                    seriesScreenTimeNonMatch,
                }
            );

            const gqlResult = await graphql({
                schema: neoSchema.schema,
                source: query,
                contextValue: { driver, driverConfig: { bookmarks: session.lastBookmark() } },
            });

            expect(gqlResult.errors).toBeFalsy();

            expect(gqlResult.data?.actors[0].actedIn).toHaveLength(2);
            expect(gqlResult.data).toEqual({
                actors: [
                    {
                        actedIn: expect.arrayContaining([
                            {
                                runtime: movieRuntime,
                                title: movieTitle,
                            },
                            {
                                episodes: seriesEpisodes,
                                title: seriesTitle,
                            },
                        ]),
                        name: actorName,
                    },
                ],
            });
        } finally {
            await session.close();
        }
    });
```

Sorry, pretty long winded! But the top-level filtering doesn't seem to be working in practice, because I receive all 4 movies and series back despite the filtering.

Can you look at either fixing this, or perhaps reducing the scope of this PR down to just the fix for issue #585?

**Comment by darrellwarde** [comment]
Hey @maxwellwynter, I'm going to close this PR due to inactivity. The bug which had been successfully fixed in this PR is about to be fixed with #701 being merged, which only leaves the filtering functionality which wasn't quite working in this PR. Feel free to reopen and have another crack at the filtering at any point. 🙂

**Pull request closed without merging**